{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0797c8e6-3c7a-454d-957b-806c539855eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "import base64\n",
    "import os\n",
    "import numpy as np\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "warnings.simplefilter('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "be0964c1-0def-4829-929f-50b0ddbe02eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load homology data from MMseqs2 results\n",
    "yih_homologs = pd.read_csv('yih_genes_mmseq.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "97a386b3-2270-4c77-b224-5859f9c22fa8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>target</th>\n",
       "      <th>identity</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gap_openings</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>target_start</th>\n",
       "      <th>target_end</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>genome_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AYG21329.1|CP032667.1</td>\n",
       "      <td>CP018976.1_cds_AQV54885.1_3704</td>\n",
       "      <td>1.000</td>\n",
       "      <td>690</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>690</td>\n",
       "      <td>1</td>\n",
       "      <td>690</td>\n",
       "      <td>1.900000e-158</td>\n",
       "      <td>489</td>\n",
       "      <td>GCA_002012045.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AYG21329.1|CP032667.1</td>\n",
       "      <td>CP018976.1_cds_AQV54885.1_3704</td>\n",
       "      <td>0.866</td>\n",
       "      <td>213</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>612</td>\n",
       "      <td>400</td>\n",
       "      <td>612</td>\n",
       "      <td>400</td>\n",
       "      <td>6.455000e-37</td>\n",
       "      <td>130</td>\n",
       "      <td>GCA_002012045.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AYG21329.1|CP032667.1</td>\n",
       "      <td>CP018976.1_cds_AQV54885.1_3704</td>\n",
       "      <td>0.975</td>\n",
       "      <td>123</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>163</td>\n",
       "      <td>41</td>\n",
       "      <td>163</td>\n",
       "      <td>1.682000e-22</td>\n",
       "      <td>87</td>\n",
       "      <td>GCA_002012045.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AYG21329.1|CP032667.1</td>\n",
       "      <td>CP018976.1_cds_AQV54885.1_3704</td>\n",
       "      <td>0.971</td>\n",
       "      <td>105</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>317</td>\n",
       "      <td>213</td>\n",
       "      <td>317</td>\n",
       "      <td>213</td>\n",
       "      <td>6.730000e-18</td>\n",
       "      <td>74</td>\n",
       "      <td>GCA_002012045.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AYG21329.1|CP032667.1</td>\n",
       "      <td>CP018976.1_cds_AQV54885.1_3704</td>\n",
       "      <td>0.917</td>\n",
       "      <td>102</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>104</td>\n",
       "      <td>3</td>\n",
       "      <td>104</td>\n",
       "      <td>3</td>\n",
       "      <td>4.892000e-16</td>\n",
       "      <td>68</td>\n",
       "      <td>GCA_002012045.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   query                          target  identity  \\\n",
       "0  AYG21329.1|CP032667.1  CP018976.1_cds_AQV54885.1_3704     1.000   \n",
       "1  AYG21329.1|CP032667.1  CP018976.1_cds_AQV54885.1_3704     0.866   \n",
       "2  AYG21329.1|CP032667.1  CP018976.1_cds_AQV54885.1_3704     0.975   \n",
       "3  AYG21329.1|CP032667.1  CP018976.1_cds_AQV54885.1_3704     0.971   \n",
       "4  AYG21329.1|CP032667.1  CP018976.1_cds_AQV54885.1_3704     0.917   \n",
       "\n",
       "   alignment_length  mismatches  gap_openings  query_start  query_end  \\\n",
       "0               690           0             0            1        690   \n",
       "1               213          28             0          612        400   \n",
       "2               123           3             0           41        163   \n",
       "3               105           3             0          317        213   \n",
       "4               102           8             0          104          3   \n",
       "\n",
       "   target_start  target_end         evalue  bit_score        genome_id  \n",
       "0             1         690  1.900000e-158        489  GCA_002012045.1  \n",
       "1           612         400   6.455000e-37        130  GCA_002012045.1  \n",
       "2            41         163   1.682000e-22         87  GCA_002012045.1  \n",
       "3           317         213   6.730000e-18         74  GCA_002012045.1  \n",
       "4           104           3   4.892000e-16         68  GCA_002012045.1  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "yih_homologs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "384c1d1d-5b79-40ac-aba3-12746fdc6962",
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_github_folder(repo, folder_path, local_dir):\n",
    "    \"\"\"\n",
    "    Download a folder from a GitHub repo\n",
    "    \"\"\"\n",
    "    api_url = f\"https://api.github.com/repos/{repo}/contents/{folder_path}\"\n",
    "    response = requests.get(api_url)\n",
    "    response.raise_for_status()\n",
    "    \n",
    "    os.makedirs(local_dir, exist_ok=True)\n",
    "    \n",
    "    for item in response.json():\n",
    "        if item['type'] == 'file':\n",
    "            file_response = requests.get(item['download_url'])\n",
    "            file_path = os.path.join(local_dir, item['name'])\n",
    "            \n",
    "            with open(file_path, 'wb') as f:\n",
    "                f.write(file_response.content)\n",
    "            print(f\"Downloaded: {item['name']}\")\n",
    "        elif item['type'] == 'dir':\n",
    "            subfolder_path = os.path.join(local_dir, item['name'])\n",
    "            download_github_folder(repo, f\"{folder_path}/{item['name']}\", subfolder_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33e49911-bf1a-4ea9-9162-3a90d752c520",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Downloads genomic data from GitHub repository (ShigellaProject)\n",
    "# If not downloaded, get genomic fna (will take some time to download) \n",
    "repo = \"zsfrbkv/ShigellaProject\"\n",
    "folder_path = \"1Tree/Data/Genomes_fna\"\n",
    "local_dir = \"Genomes_fna\"\n",
    "download_github_folder(repo, folder_path, local_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe159ab0-c178-46c8-a840-6bc7aeac3a66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extracts genomic coordinates and strand information from FASTA descriptions\n",
    "# Create mapping target2info (will take some time to pase fasta files)\n",
    "fna_dir = local_dir\n",
    "genome_id_lst = yih_homologs.genome_id.unique()\n",
    "import glob\n",
    "target2info = {}\n",
    "for genome_id in genome_id_lst:\n",
    "    fna_pattern = os.path.join(fna_dir, f'{genome_id}*.fna')\n",
    "    fna_files = glob.glob(fna_pattern)\n",
    "    \n",
    "    if fna_files:\n",
    "        fna_f = fna_files[0]\n",
    "        target2info.update({r.id.replace('lcl|', ''): r.description\n",
    "              for r in SeqIO.parse(fna_f, 'fasta')})\n",
    "    else:\n",
    "        print(f\"Warning: No .fna file found for genome_id: {genome_id}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "558d7053-ee30-40a8-bbaf-65b0a8bf0768",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Maps protein IDs to gene names \n",
    "protein_queryGene = dict(zip(['AYG21329.1|CP032667.1', 'AYG21328.1|CP032667.1',\n",
    "       'AYG21327.1|CP032667.1', 'AYG21326.1|CP032667.1',\n",
    "       'AYG21325.1|CP032667.1', 'AYG21324.1|CP032667.1',\n",
    "       'AYG21323.1|CP032667.1', 'AYG21322.1|CP032667.1',\n",
    "       'AYG21321.1|CP032667.1', 'AYG21320.1|CP032667.1'],\n",
    "        ['ompL', 'yihO', 'yihP', 'yihQ', \n",
    "        'yihR', 'yihS', 'yihT', 'yihU', 'yihV', 'yihW'])) \n",
    "\n",
    "yih_homologs['query_gene'] = yih_homologs['query'].map(protein_queryGene)\n",
    "yih_homologs['target_info'] = yih_homologs.target.map(target2info)\n",
    "yih_homologs['target_location'] = yih_homologs.target_info.str.extract(r'\\[location=([^\\]]+)\\]')\n",
    "yih_homologs['strand'] = yih_homologs['target_location'].apply(lambda x: '-' if 'complement' in x else '+')\n",
    "yih_homologs[['start', 'end']] = yih_homologs['target_location'].\\\n",
    "apply(lambda x: x.replace('complement(', '').replace(')', '')).\\\n",
    "astype(str).str.split('\\.\\.', expand=True).astype(int)\n",
    "\n",
    "yih_homologs = yih_homologs[yih_homologs.groupby(['target', 'query'])['bit_score'].\\\n",
    "transform(max) == yih_homologs['bit_score']]\n",
    "yih_homologs['locus_tag'] = yih_homologs.target_info.str.extract(r'\\[locus_tag=([^\\]]+)\\]')\n",
    "yih_homologs['chr_id'] = yih_homologs.target.str.split('_', expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "11c41268-2da1-43a8-8907-e412bfdb9081",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_queryGene = dict(zip(['AYG21329.1|CP032667.1', 'AYG21328.1|CP032667.1',\n",
    "       'AYG21327.1|CP032667.1', 'AYG21326.1|CP032667.1',\n",
    "       'AYG21325.1|CP032667.1', 'AYG21324.1|CP032667.1',\n",
    "       'AYG21323.1|CP032667.1', 'AYG21322.1|CP032667.1',\n",
    "       'AYG21321.1|CP032667.1', 'AYG21320.1|CP032667.1'],\n",
    "        ['ompL', 'yihO', 'yihP', 'yihQ', \n",
    "        'yihR', 'yihS', 'yihT', 'yihU', 'yihV', 'yihW'])) \n",
    "\n",
    "def get_colocalization(blastp_drop,\n",
    "                       map_query_gene=protein_queryGene, \n",
    "                       threshold=5000):\n",
    "    # blastp_drop = pd.read_csv(args.blastpOutputExt, sep='\\t')\n",
    "    # blastp_drop['query_gene'] = blastp_drop.query_protein.map(map_query_gene)\n",
    "\n",
    "    # Compute midpoint coordinates\n",
    "    blastp_drop['midpoint'] = (blastp_drop.start + blastp_drop.end) / 2\n",
    "    \n",
    "    # Sort by chromosome and midpoint\n",
    "    blastp_drop_sorted = blastp_drop.groupby(\"chr_id\").apply(\n",
    "        lambda x: x.sort_values(\"midpoint\")).reset_index(drop=True)\n",
    "    \n",
    "    # Find locus composition\n",
    "    dict_assembly_queryGenes = {}\n",
    "    dict_assembly_locusTags = {}\n",
    "    dict_assembly_coords = {}\n",
    "    \n",
    "    for chr_id, group in blastp_drop_sorted.groupby(\"chr_id\"):\n",
    "        midpoints = group.midpoint.values\n",
    "        locusTags = group.locus_tag.values\n",
    "        queryGenes = group.query_gene.values\n",
    "        \n",
    "        # Initialize first locus\n",
    "        current_loci = {\n",
    "            'queryGenes': [queryGenes[0]],\n",
    "            'locusTags': [str(locusTags[0])],\n",
    "            'coords': [str(midpoints[0])]\n",
    "        }\n",
    "        \n",
    "        queryGenes_loci = []\n",
    "        locusTags_loci = []\n",
    "        coords_loci = []\n",
    "        \n",
    "        for k in range(len(midpoints) - 1):\n",
    "            if (midpoints[k + 1] - midpoints[k]) <= threshold:\n",
    "                # Add to current locus\n",
    "                current_loci['queryGenes'].append(queryGenes[k + 1])\n",
    "                current_loci['locusTags'].append(str(locusTags[k + 1]))\n",
    "                current_loci['coords'].append(str(midpoints[k + 1]))\n",
    "            else:\n",
    "                # Save current locus and start new one\n",
    "                queryGenes_loci.append('/'.join(current_loci['queryGenes']))\n",
    "                locusTags_loci.append('/'.join(current_loci['locusTags']))\n",
    "                coords_loci.append('/'.join(current_loci['coords']))\n",
    "                \n",
    "                # Start new locus\n",
    "                current_loci = {\n",
    "                    'queryGenes': [queryGenes[k + 1]],\n",
    "                    'locusTags': [str(locusTags[k + 1])],\n",
    "                    'coords': [str(midpoints[k + 1])]\n",
    "                }\n",
    "        \n",
    "        # Add the last locus\n",
    "        queryGenes_loci.append('/'.join(current_loci['queryGenes']))\n",
    "        locusTags_loci.append('/'.join(current_loci['locusTags']))\n",
    "        coords_loci.append('/'.join(current_loci['coords']))\n",
    "        \n",
    "        # Join all loci for this chromosome\n",
    "        dict_assembly_queryGenes[chr_id] = ' '.join(queryGenes_loci)\n",
    "        dict_assembly_locusTags[chr_id] = ' '.join(locusTags_loci)\n",
    "        dict_assembly_coords[chr_id] = ' '.join(coords_loci)\n",
    "    \n",
    "    # Create mapping dictionaries\n",
    "    queryGenes_list = list(map_query_gene.values())\n",
    "    dict_locusTag_queryGeneBlock = {}\n",
    "    dict_locusTag_locusTagBlock = {}\n",
    "    \n",
    "    for chr_id in blastp_drop_sorted.chr_id.unique():\n",
    "        locusTag_blocks = dict_assembly_locusTags[chr_id].split()\n",
    "        gene_blocks = dict_assembly_queryGenes[chr_id].split()\n",
    "        \n",
    "        for locusTag_block, gene_block in zip(locusTag_blocks, gene_blocks):\n",
    "            locus_tags = locusTag_block.split('/')\n",
    "            genes = gene_block.split('/')\n",
    "            \n",
    "            # Determine order based on queryGenes_list\n",
    "            if (queryGenes_list.index(genes[0]) > \n",
    "                queryGenes_list.index(genes[-1])):\n",
    "                gene_block = '/'.join(reversed(genes))\n",
    "                locusTag_block = '/'.join(reversed(locus_tags))\n",
    "            \n",
    "            for locus_tag in locus_tags:\n",
    "                dict_locusTag_queryGeneBlock[locus_tag] = gene_block\n",
    "                dict_locusTag_locusTagBlock[locus_tag] = locusTag_block\n",
    "    \n",
    "    # Add locus information columns\n",
    "    blastp_drop_sorted['queryGenes_locus'] = blastp_drop_sorted.locus_tag.map(\n",
    "        dict_locusTag_queryGeneBlock)\n",
    "    blastp_drop_sorted['locusTag_locus'] = blastp_drop_sorted.locus_tag.map(\n",
    "        dict_locusTag_locusTagBlock)\n",
    "    return blastp_drop_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "648a577c-b343-4ea7-8b93-c7453b490b5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Will take some time\n",
    "# Identifies colocalized genes using a proximity threshold (5kb default)\n",
    "yih_homologs_loci = get_colocalization(yih_homologs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ff5dc72-c58c-4a5f-829e-89cd00f1ffc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "yih_homologs_loci['set_queryGenes_locus'] = yih_homologs_loci.\\\n",
    "queryGenes_locus.apply(lambda x: '~'.join(sorted(set(x.split('/')))))\n",
    "yih_homologs_loci = yih_homologs_loci[yih_homologs_loci.set_queryGenes_locus.apply(lambda x: 'yihV' in x and 'yihT' in x)]\n",
    "yih_homologs_loci = yih_homologs_loci.drop_duplicates('target')\n",
    "\n",
    "yih_homologs_loci['start_min'] = yih_homologs_loci.groupby('chr_id')['start'].transform(min)\n",
    "yih_homologs_loci['end_max'] = yih_homologs_loci.groupby('chr_id')['end'].transform(max)\n",
    "yih_homologs_loci['set_locusTag_locus'] = yih_homologs_loci['locusTag_locus'].\\\n",
    "apply(lambda x: '~'.join(sorted(set(x.split('/')))))\n",
    "\n",
    "# Classifies operon variants: \"long\" variant: contains yihQ or yihS genes\n",
    "# \"short\" variant: missing these genes\n",
    "yih_homologs_loci['variant'] = yih_homologs_loci.set_queryGenes_locus.apply(lambda x: \n",
    "                                             'long' if 'yihQ' in x or 'yihS' in x else 'short')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bd6f919-006c-4606-81f0-5262d263ee28",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save to dataframe\n",
    "## yih_homologs_loci.to_csv('yih_homologs_loci.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5342384a-7cf8-403e-a155-7261cb668080",
   "metadata": {},
   "outputs": [],
   "source": [
    "## ======= Create target2info data ========#\n",
    "target2info_data = pd.DataFrame({'target': [k for k in target2info],\n",
    "             'target_info': [target2info[k] for k in target2info]})\n",
    "\n",
    "target2info_data['target_location'] = target2info_data.target_info.str.extract(r'\\[location=([^\\]]+)\\]')\n",
    "target2info_data['strand'] = target2info_data['target_location'].apply(lambda x: '-' if 'complement' in x else '+')\n",
    "\n",
    "target2info_data = target2info_data[~target2info_data.target_location.str.contains('join')]\n",
    "target2info_data = target2info_data[~target2info_data.target_location.str.contains('<')]\n",
    "target2info_data = target2info_data[~target2info_data.target_location.str.contains('>')]\n",
    "\n",
    "\n",
    "target2info_data[['start', 'end']] = target2info_data['target_location'].\\\n",
    "apply(lambda x: x.replace('complement(', '').replace(')', '')).\\\n",
    "astype(str).str.split('\\.\\.', expand=True).astype(int)\n",
    "\n",
    "\n",
    "target2info_data['locus_tag'] = target2info_data.target_info.str.extract(r'\\[locus_tag=([^\\]]+)\\]')\n",
    "\n",
    "target2info_data['chr_id'] = target2info_data.target.str.split('_', expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03d068fd-0c22-40a6-a145-a9156e443d65",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save the dataframe\n",
    "#target2info_data.to_csv('target2info_data_altToGFFs.csv', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42c4f0aa-ca15-4917-b56a-df6bf9ad43bf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06d6c6c6-9970-4b7d-99c3-b5e62ef91d63",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
