{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7656de05-521a-40a5-bdcc-eba0c509df7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "warnings.simplefilter('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d92b1b63-dbd5-4eba-8baf-d6e9dc5c38fb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d5eb75c3-fbc6-4703-abaf-782fc1a97a6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load metadata from previous analysis \n",
    "yih_homologs_loci = pd.read_csv('yih_homologs_loci.csv', \n",
    "                                sep='\\t')\n",
    "\n",
    "target2info_data = pd.read_csv('target2info_data_altToGFFs.csv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "474b84a4-4386-4868-b026-c15e29537862",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>target_info</th>\n",
       "      <th>target_location</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>chr_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP018976.1_cds_AQV50030.1_2</td>\n",
       "      <td>lcl|CP018976.1_cds_AQV50030.1_2 [locus_tag=BE9...</td>\n",
       "      <td>1..1497</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>1497</td>\n",
       "      <td>BE949_01660</td>\n",
       "      <td>CP018976.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP018976.1_cds_AQV50031.1_3</td>\n",
       "      <td>lcl|CP018976.1_cds_AQV50031.1_3 [locus_tag=BE9...</td>\n",
       "      <td>1497..1853</td>\n",
       "      <td>+</td>\n",
       "      <td>1497</td>\n",
       "      <td>1853</td>\n",
       "      <td>BE949_01665</td>\n",
       "      <td>CP018976.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP018976.1_cds_AQV50032.1_4</td>\n",
       "      <td>lcl|CP018976.1_cds_AQV50032.1_4 [locus_tag=BE9...</td>\n",
       "      <td>1853..2122</td>\n",
       "      <td>+</td>\n",
       "      <td>1853</td>\n",
       "      <td>2122</td>\n",
       "      <td>BE949_01670</td>\n",
       "      <td>CP018976.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP018976.1_cds_AQV50033.1_5</td>\n",
       "      <td>lcl|CP018976.1_cds_AQV50033.1_5 [locus_tag=BE9...</td>\n",
       "      <td>2089..2271</td>\n",
       "      <td>+</td>\n",
       "      <td>2089</td>\n",
       "      <td>2271</td>\n",
       "      <td>BE949_01675</td>\n",
       "      <td>CP018976.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP018976.1_cds_AQV50034.1_6</td>\n",
       "      <td>lcl|CP018976.1_cds_AQV50034.1_6 [locus_tag=BE9...</td>\n",
       "      <td>2264..4099</td>\n",
       "      <td>+</td>\n",
       "      <td>2264</td>\n",
       "      <td>4099</td>\n",
       "      <td>BE949_01680</td>\n",
       "      <td>CP018976.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        target  \\\n",
       "0  CP018976.1_cds_AQV50030.1_2   \n",
       "1  CP018976.1_cds_AQV50031.1_3   \n",
       "2  CP018976.1_cds_AQV50032.1_4   \n",
       "3  CP018976.1_cds_AQV50033.1_5   \n",
       "4  CP018976.1_cds_AQV50034.1_6   \n",
       "\n",
       "                                         target_info target_location strand  \\\n",
       "0  lcl|CP018976.1_cds_AQV50030.1_2 [locus_tag=BE9...         1..1497      +   \n",
       "1  lcl|CP018976.1_cds_AQV50031.1_3 [locus_tag=BE9...      1497..1853      +   \n",
       "2  lcl|CP018976.1_cds_AQV50032.1_4 [locus_tag=BE9...      1853..2122      +   \n",
       "3  lcl|CP018976.1_cds_AQV50033.1_5 [locus_tag=BE9...      2089..2271      +   \n",
       "4  lcl|CP018976.1_cds_AQV50034.1_6 [locus_tag=BE9...      2264..4099      +   \n",
       "\n",
       "   start   end    locus_tag      chr_id  \n",
       "0      1  1497  BE949_01660  CP018976.1  \n",
       "1   1497  1853  BE949_01665  CP018976.1  \n",
       "2   1853  2122  BE949_01670  CP018976.1  \n",
       "3   2089  2271  BE949_01675  CP018976.1  \n",
       "4   2264  4099  BE949_01680  CP018976.1  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target2info_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62f85ae4-997c-4e5a-b542-60e15ceadc85",
   "metadata": {},
   "source": [
    "# Generate annotation file that schematically plots _yih_ gene cassettes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 410,
   "id": "f2118391-a285-4dba-8bd7-f8959966a342",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_domains(strand, start, mid, gene, color, end):\n",
    "    \"\"\"Generate iTOL-style domain arrows based on strand.\"\"\"\n",
    "    if strand == \"-\":\n",
    "        return f\"TL|{start}|{mid}|{color}|,RE|{mid}|{end}|{color}|{gene}\"\n",
    "    return f\"RE|{start}|{mid}|{color}|{gene},TR|{mid}|{end}|{color}|\"\n",
    "\n",
    "\n",
    "def flip_orientation_locus(tree_data, ref_gene, ref_strand, show_ID=True):\n",
    "    #tree_data['length'] = tree_data.chr_id.map(tree_data.groupby('chr_id').agg({'start':min, 'end':max}).diff(axis=1).to_dict()['end'])\n",
    "    tree_data['length'] = tree_data.locusTag_locus.map(tree_data.groupby('locusTag_locus').\\\n",
    "                                                       agg({'start':min, 'end':max}).\\\n",
    "                                                       diff(axis=1).to_dict()['end'])\n",
    "\n",
    "    locusTag_locus_F = tree_data[(tree_data.gene == ref_gene) & \\\n",
    "                                 (tree_data.strand == ref_strand)].locusTag_locus.unique()\n",
    "    locusTag_locus_R = tree_data[(tree_data.gene == ref_gene) & \\\n",
    "                                 (tree_data.strand != ref_strand)].locusTag_locus.unique()\n",
    "    \n",
    "    tree_data_F = tree_data[tree_data.locusTag_locus.isin(locusTag_locus_F)]\n",
    "    tree_data_R = tree_data[tree_data.locusTag_locus.isin(locusTag_locus_R)]\n",
    "    \n",
    "    # forward data\n",
    "    tree_data_F['coord_2subtract'] = tree_data_F.locusTag_locus.map(tree_data_F.groupby('locusTag_locus').\\\n",
    "                                                                    agg({'start': min}).to_dict()['start']) - 1\n",
    "\n",
    "    tree_data_F.start = tree_data_F.start - tree_data_F.coord_2subtract\n",
    "    tree_data_F.end = tree_data_F.end - tree_data_F.coord_2subtract\n",
    "    \n",
    "    \n",
    "    tree_data_F[\"mid_coord\"] = (round(400 + tree_data_F.start)).astype(int)\n",
    "    if show_ID:\n",
    "        tree_data_F['domains'] = tree_data_F.apply(lambda x: get_domains(x.strand, x.start, x.mid_coord, x.query_gene + ' '+ str(round(x.identity,1)), x.color, x.end),  axis=1)\n",
    "    else:\n",
    "        tree_data_F['domains'] = tree_data_F.apply(lambda x: get_domains(x.strand, x.start, x.mid_coord, x.gene, x.color, x.end),  axis=1)\n",
    "\n",
    "    # reverse data\n",
    "    tree_data_R['coord_2add'] = tree_data_R.locusTag_locus.map(tree_data_R.groupby('locusTag_locus').agg({'end': max}).to_dict()['end']) + 1\n",
    "    \n",
    "    tree_data_R.start = -tree_data_R.start\n",
    "    tree_data_R.end = -tree_data_R.end\n",
    "\n",
    "    tree_data_R.start = tree_data_R.start + tree_data_R.coord_2add\n",
    "    tree_data_R.end = tree_data_R.end + tree_data_R.coord_2add\n",
    "    tree_data_R = tree_data_R.sort_values(['locusTag_locus', 'start'])\n",
    "    tree_data_R[\"strand\"] = tree_data_R.strand.map({'-': \"+\", '+': '-'})\n",
    "    tree_data_R = tree_data_R.rename(columns={'start':'end', 'end':'start'})\n",
    "    \n",
    "    tree_data_R[\"mid_coord\"] = (round(400 + tree_data_R.start)).astype(int)\n",
    "    if show_ID:\n",
    "        tree_data_R['domains'] = tree_data_R.apply(lambda x:\n",
    "                                                   get_domains(x.strand, x.start, x.mid_coord, x.query_gene + ' '+ str(round(x.identity,1)), x.color, x.end),  axis=1)\n",
    "    else:\n",
    "        tree_data_R['domains'] = tree_data_R.apply(lambda x:\n",
    "                                                   get_domains(x.strand, x.start, x.mid_coord, x.gene, x.color, x.end),  axis=1)\n",
    "\n",
    "    \n",
    "    tree_data_FR = pd.concat([tree_data_R, tree_data_F])\n",
    "    \n",
    "    node_domains_map = tree_data_FR.groupby('locusTag_locus').agg({'domains': lambda x: ','.join(x)}).to_dict()['domains'] \n",
    "\n",
    "    \n",
    "    tree_data_FR['all_domains'] = tree_data_FR.locusTag_locus.map(node_domains_map)\n",
    "\n",
    "    ### IMPORTANT \n",
    "    tree_data_FR['line'] = tree_data_FR.terminal_node +',' + (tree_data_FR.length + 1).astype(str) + ',' + tree_data_FR.all_domains\n",
    "    \n",
    "    max_length = max(tree_data_FR.length)\n",
    "    return tree_data_FR\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def write_itol_domains(df_all, out_path, annotate_colors):\n",
    "    with open(out_path, 'w') as f:\n",
    "        f.write(\"DATASET_DOMAINS\\nSEPARATOR COMMA\\n\")\n",
    "        f.write(\"DATASET_LABEL,loci\\n\")\n",
    "        f.write(\"LEGEND_TITLE,genes\\n\")\n",
    "        f.write(\"LEGEND_SHAPES,\" + \",\".join(['RE'] * len(annotate_colors)) + \"\\n\")\n",
    "        f.write(\"LEGEND_COLORS,\" + \",\".join([annotate_colors[g] \\\n",
    "                                             for g in ['ompL', 'yihO', 'yihP',\n",
    "                                                       'yihQ', 'yihR', 'yihS',\n",
    "                                                       'yihT', 'yihU', 'yihV',\n",
    "                                                       'yihW', 'other']]) + \"\\n\")\n",
    "        f.write(\"LEGEND_LABELS,\" + \",\".join(['ompL (porin)',\n",
    "                                             'yihO (MFS transporter)',\n",
    "                                             'yihP (MFS transporter)',\n",
    "                                             'yihQ (SQase)',\n",
    "                                             'yihR (mutarotase)',\n",
    "                                             'yihS (isomerase)',\n",
    "                                             'yihT (aldolase)',\n",
    "                                             'yihU (reductase)',\n",
    "                                             'yihV (kinase)',\n",
    "                                             'yihW (csqR) (regulator)', \n",
    "                                             'other']) + \"\\n\")\n",
    "        f.write(\"BACKBONE_COLOR,#ffffff\\nSHOW_DOMAIN_LABELS,1\\nBORDER_WIDTH,0\\nDATA\\n\")\n",
    "        for line in df_all['line'].dropna().unique():\n",
    "            f.write(line + '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "69179f99-d8ed-45ac-b3ff-c42cb62fa5bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotate_colors = {\n",
    "    'yihU': '#00008B',   # very dark blue\n",
    "    'yihT': '#1A2C5F',   # dark desaturated blue\n",
    "    'yihV': '#4343F7',   # bright blue\n",
    "    'yihS': '#6683E9',   # medium blue\n",
    "    'yihR': '#99CCFF',   # soft sky blue\n",
    "    'yihQ': '#CCE5FF',   # very light blue\n",
    "\n",
    "    'yihW': '#708090',  # DarkSlateBlue\n",
    "    'ompL': '#E5E4E2',   # light gray\n",
    "    'yihO': '#C0C0C0',   # classic gray\n",
    "    'yihP':  '#C0C0C0', #'#848884',   # gray-green\n",
    "\n",
    "    'other': '#F9F6EE'   # fallback color (light cream/neutral)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "551c3df5-a87c-497f-aac0-bf6d1d3d5f19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare data on yih loci \n",
    "# filters genes within yih genomic ranges\n",
    "chr_ranges = yih_homologs_loci.\\\n",
    "drop_duplicates(['chr_id',  'start_min',  'end_max'])[['chr_id', 'start_min',\\\n",
    "                                                       'end_max',  'set_locusTag_locus', \\\n",
    "                                                       'locusTag_locus', 'set_queryGenes_locus']]\n",
    "\n",
    "# merge with target2info_data to filter genes within these ranges\n",
    "subset_target2info = pd.merge(\n",
    "    target2info_data,\n",
    "    chr_ranges,\n",
    "    on='chr_id',\n",
    "    how='inner'\n",
    ").query('start >= start_min and end <= end_max')\n",
    "\n",
    "# drop the temporary range columns if needed\n",
    "subset_target2info = subset_target2info.drop(['start_min', 'end_max'], axis=1)\n",
    "\n",
    "# sort by chromosome and position\n",
    "subset_target2info = subset_target2info.sort_values(['chr_id', 'start'])\n",
    "\n",
    "subset_target2info = pd.merge(\n",
    "    subset_target2info,\n",
    "    yih_homologs[yih_homologs.groupby(['target', 'query'])['bit_score']\\\n",
    "    .transform(max) == yih_homologs['bit_score']][['locus_tag', 'genome_id', 'query_gene']],\n",
    "    on='locus_tag',\n",
    "    how='left'\n",
    ").rename(columns={'genome_id': 'terminal_node'})\n",
    "\n",
    "subset_target2info['gene'] = subset_target2info['query_gene'].fillna('other')\n",
    "subset_target2info['color'] = subset_target2info.gene.map(annotate_colors)\n",
    "subset_target2info = subset_target2info.drop_duplicates('locus_tag')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a29bebc-74fc-4361-aff9-1704c5f9a2fd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d8fbb42-a83e-46f3-aa8d-7a0cd93813ac",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Standardizes orientations using yihW as reference gene\n",
    "df_prepared = flip_orientation_locus(subset_target2info, \n",
    "                                     ref_gene='yihW',\n",
    "                                     ref_strand='+', show_ID=False)\n",
    "\n",
    "# Generate domain coordinates for arrow visualization\n",
    "write_itol_domains(df_prepared, 'Ecoli.domains.txt', annotate_colors)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15ffbc25-68d0-438d-b3a9-f236284b79a0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b76cb622-ca93-4970-ab6e-4bd74a3399e5",
   "metadata": {},
   "source": [
    "# Generate annotation file describing _yih_ gene cassette content in a binary format (gene presence/absence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7f4ebece-cfb3-4f8a-a212-82f15110ecd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "yih_homologs_loci = pd.read_csv('yih_homologs_loci.csv', \n",
    "                                sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b438be21-a0e9-402b-b52f-5ad55604a709",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>target</th>\n",
       "      <th>identity</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gap_openings</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>target_start</th>\n",
       "      <th>target_end</th>\n",
       "      <th>...</th>\n",
       "      <th>midpoint</th>\n",
       "      <th>queryGenes_locus</th>\n",
       "      <th>locusTag_locus</th>\n",
       "      <th>set_queryGenes_locus</th>\n",
       "      <th>start_min</th>\n",
       "      <th>end_max</th>\n",
       "      <th>set_locusTag_locus</th>\n",
       "      <th>variant</th>\n",
       "      <th>Info</th>\n",
       "      <th>Cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AYG21329.1|CP032667.1</td>\n",
       "      <td>AE005674.2_cds_AAN45382.1_3949</td>\n",
       "      <td>0.961</td>\n",
       "      <td>330</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>330</td>\n",
       "      <td>37</td>\n",
       "      <td>366</td>\n",
       "      <td>...</td>\n",
       "      <td>4076793.0</td>\n",
       "      <td>ompL/yihP/yihO/yihO/yihP/yihQ/yihR/yihS/yihT/y...</td>\n",
       "      <td>SF3947/SF3948/SF3948/SF3949/SF3949/SF3950/SF39...</td>\n",
       "      <td>ompL~yihO~yihP~yihQ~yihR~yihS~yihT~yihU~yihV~yihW</td>\n",
       "      <td>4076600</td>\n",
       "      <td>4088098</td>\n",
       "      <td>SF3947~SF3948~SF3949~SF3950~SF3951~SF3952~SF39...</td>\n",
       "      <td>long</td>\n",
       "      <td>shigellosis</td>\n",
       "      <td>7 (B1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AYG21327.1|CP032667.1</td>\n",
       "      <td>AE005674.2_cds_AAN45383.2_3950</td>\n",
       "      <td>0.686</td>\n",
       "      <td>1365</td>\n",
       "      <td>425</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>1374</td>\n",
       "      <td>1</td>\n",
       "      <td>1356</td>\n",
       "      <td>...</td>\n",
       "      <td>4077719.5</td>\n",
       "      <td>ompL/yihP/yihO/yihO/yihP/yihQ/yihR/yihS/yihT/y...</td>\n",
       "      <td>SF3947/SF3948/SF3948/SF3949/SF3949/SF3950/SF39...</td>\n",
       "      <td>ompL~yihO~yihP~yihQ~yihR~yihS~yihT~yihU~yihV~yihW</td>\n",
       "      <td>4076600</td>\n",
       "      <td>4088098</td>\n",
       "      <td>SF3947~SF3948~SF3949~SF3950~SF3951~SF3952~SF39...</td>\n",
       "      <td>long</td>\n",
       "      <td>shigellosis</td>\n",
       "      <td>7 (B1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AYG21328.1|CP032667.1</td>\n",
       "      <td>AE005674.2_cds_AAN45384.1_3951</td>\n",
       "      <td>0.659</td>\n",
       "      <td>1029</td>\n",
       "      <td>347</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1020</td>\n",
       "      <td>43</td>\n",
       "      <td>1071</td>\n",
       "      <td>...</td>\n",
       "      <td>4079244.5</td>\n",
       "      <td>ompL/yihP/yihO/yihO/yihP/yihQ/yihR/yihS/yihT/y...</td>\n",
       "      <td>SF3947/SF3948/SF3948/SF3949/SF3949/SF3950/SF39...</td>\n",
       "      <td>ompL~yihO~yihP~yihQ~yihR~yihS~yihT~yihU~yihV~yihW</td>\n",
       "      <td>4076600</td>\n",
       "      <td>4088098</td>\n",
       "      <td>SF3947~SF3948~SF3949~SF3950~SF3951~SF3952~SF39...</td>\n",
       "      <td>long</td>\n",
       "      <td>shigellosis</td>\n",
       "      <td>7 (B1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AYG21326.1|CP032667.1</td>\n",
       "      <td>AE005674.2_cds_AAN45385.1_3952</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2034</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2034</td>\n",
       "      <td>1</td>\n",
       "      <td>2034</td>\n",
       "      <td>...</td>\n",
       "      <td>4080902.0</td>\n",
       "      <td>ompL/yihP/yihO/yihO/yihP/yihQ/yihR/yihS/yihT/y...</td>\n",
       "      <td>SF3947/SF3948/SF3948/SF3949/SF3949/SF3950/SF39...</td>\n",
       "      <td>ompL~yihO~yihP~yihQ~yihR~yihS~yihT~yihU~yihV~yihW</td>\n",
       "      <td>4076600</td>\n",
       "      <td>4088098</td>\n",
       "      <td>SF3947~SF3948~SF3949~SF3950~SF3951~SF3952~SF39...</td>\n",
       "      <td>long</td>\n",
       "      <td>shigellosis</td>\n",
       "      <td>7 (B1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AYG21325.1|CP032667.1</td>\n",
       "      <td>AE005674.2_cds_AAN45386.2_3953</td>\n",
       "      <td>0.986</td>\n",
       "      <td>900</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>924</td>\n",
       "      <td>1</td>\n",
       "      <td>900</td>\n",
       "      <td>...</td>\n",
       "      <td>4082570.0</td>\n",
       "      <td>ompL/yihP/yihO/yihO/yihP/yihQ/yihR/yihS/yihT/y...</td>\n",
       "      <td>SF3947/SF3948/SF3948/SF3949/SF3949/SF3950/SF39...</td>\n",
       "      <td>ompL~yihO~yihP~yihQ~yihR~yihS~yihT~yihU~yihV~yihW</td>\n",
       "      <td>4076600</td>\n",
       "      <td>4088098</td>\n",
       "      <td>SF3947~SF3948~SF3949~SF3950~SF3951~SF3952~SF39...</td>\n",
       "      <td>long</td>\n",
       "      <td>shigellosis</td>\n",
       "      <td>7 (B1)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   query                          target  identity  \\\n",
       "0  AYG21329.1|CP032667.1  AE005674.2_cds_AAN45382.1_3949     0.961   \n",
       "1  AYG21327.1|CP032667.1  AE005674.2_cds_AAN45383.2_3950     0.686   \n",
       "2  AYG21328.1|CP032667.1  AE005674.2_cds_AAN45384.1_3951     0.659   \n",
       "3  AYG21326.1|CP032667.1  AE005674.2_cds_AAN45385.1_3952     1.000   \n",
       "4  AYG21325.1|CP032667.1  AE005674.2_cds_AAN45386.2_3953     0.986   \n",
       "\n",
       "   alignment_length  mismatches  gap_openings  query_start  query_end  \\\n",
       "0               330          13             0            1        330   \n",
       "1              1365         425             0           10       1374   \n",
       "2              1029         347             0            1       1020   \n",
       "3              2034           0             0            1       2034   \n",
       "4               900          13             0           25        924   \n",
       "\n",
       "   target_start  target_end  ...   midpoint  \\\n",
       "0            37         366  ...  4076793.0   \n",
       "1             1        1356  ...  4077719.5   \n",
       "2            43        1071  ...  4079244.5   \n",
       "3             1        2034  ...  4080902.0   \n",
       "4             1         900  ...  4082570.0   \n",
       "\n",
       "                                    queryGenes_locus  \\\n",
       "0  ompL/yihP/yihO/yihO/yihP/yihQ/yihR/yihS/yihT/y...   \n",
       "1  ompL/yihP/yihO/yihO/yihP/yihQ/yihR/yihS/yihT/y...   \n",
       "2  ompL/yihP/yihO/yihO/yihP/yihQ/yihR/yihS/yihT/y...   \n",
       "3  ompL/yihP/yihO/yihO/yihP/yihQ/yihR/yihS/yihT/y...   \n",
       "4  ompL/yihP/yihO/yihO/yihP/yihQ/yihR/yihS/yihT/y...   \n",
       "\n",
       "                                      locusTag_locus  \\\n",
       "0  SF3947/SF3948/SF3948/SF3949/SF3949/SF3950/SF39...   \n",
       "1  SF3947/SF3948/SF3948/SF3949/SF3949/SF3950/SF39...   \n",
       "2  SF3947/SF3948/SF3948/SF3949/SF3949/SF3950/SF39...   \n",
       "3  SF3947/SF3948/SF3948/SF3949/SF3949/SF3950/SF39...   \n",
       "4  SF3947/SF3948/SF3948/SF3949/SF3949/SF3950/SF39...   \n",
       "\n",
       "                                set_queryGenes_locus start_min  end_max  \\\n",
       "0  ompL~yihO~yihP~yihQ~yihR~yihS~yihT~yihU~yihV~yihW   4076600  4088098   \n",
       "1  ompL~yihO~yihP~yihQ~yihR~yihS~yihT~yihU~yihV~yihW   4076600  4088098   \n",
       "2  ompL~yihO~yihP~yihQ~yihR~yihS~yihT~yihU~yihV~yihW   4076600  4088098   \n",
       "3  ompL~yihO~yihP~yihQ~yihR~yihS~yihT~yihU~yihV~yihW   4076600  4088098   \n",
       "4  ompL~yihO~yihP~yihQ~yihR~yihS~yihT~yihU~yihV~yihW   4076600  4088098   \n",
       "\n",
       "                                  set_locusTag_locus  variant         Info  \\\n",
       "0  SF3947~SF3948~SF3949~SF3950~SF3951~SF3952~SF39...     long  shigellosis   \n",
       "1  SF3947~SF3948~SF3949~SF3950~SF3951~SF3952~SF39...     long  shigellosis   \n",
       "2  SF3947~SF3948~SF3949~SF3950~SF3951~SF3952~SF39...     long  shigellosis   \n",
       "3  SF3947~SF3948~SF3949~SF3950~SF3951~SF3952~SF39...     long  shigellosis   \n",
       "4  SF3947~SF3948~SF3949~SF3950~SF3951~SF3952~SF39...     long  shigellosis   \n",
       "\n",
       "  Cluster  \n",
       "0  7 (B1)  \n",
       "1  7 (B1)  \n",
       "2  7 (B1)  \n",
       "3  7 (B1)  \n",
       "4  7 (B1)  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "yih_homologs_loci.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "67746913-f55c-43e5-b4b9-d87c61ce35bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create binary matrix of gene presence/absence\n",
    "\n",
    "crosstab = pd.crosstab(yih_homologs_loci['genome_id'],\n",
    "                       yih_homologs_loci['query_gene'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "58e549ab-cb60-49bd-b8d1-5013220abad4",
   "metadata": {},
   "outputs": [],
   "source": [
    "crosstab.loc[crosstab['yihP'] == 2, 'yihO'] = 1\n",
    "crosstab.loc[crosstab['yihO'] == 2, 'yihP'] = 1\n",
    "crosstab  = crosstab.replace(2, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8fce101f-a9e8-443d-8664-307d497f7f67",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene2color = {'yihU': '#00008B', 'yihT': '#1A2C5F',\n",
    "              'yihV': '#4343f7', 'yihS': '#6683E9',\n",
    "              'yihR': '#99CCFF', 'yihQ': '#CCE5FF',\n",
    "              'ompL': '#E5E4E2', 'yihO': '#C0C0C0',\n",
    "              'yihP': '#C0C0C0', 'yihW': '#708090'}\n",
    "genes_order = ['ompL', 'yihO', 'yihP', 'yihQ',\n",
    "               'yihR', 'yihS', 'yihT', 'yihU',\n",
    "               'yihV', 'yihW']\n",
    "\n",
    "with open('yih_loci_presence.txt', 'w') as f:\n",
    "    f.write('\\n'.join(['DATASET_BINARY', 'SEPARATOR COMMA',\n",
    "          'DATASET_LABEL,genes', 'COLOR,#ff0000', \n",
    "                       f'FIELD_SHAPES,{\",\".join([\"2\"]*len(genes_order))}'])+'\\n')\n",
    "\n",
    "    f.write(f'LEGEND_TITLE,genes\\n')\n",
    "    f.write(f'LEGEND_SHAPES,{\",\".join([\"2\"]*len(genes_order))}\\n')\n",
    "    f.write(f'LEGEND_SHAPES,{\",\".join([\"2\"]*len(genes_order))}\\n')\n",
    "    f.write(f'LEGEND_LABELS,{\",\".join([i for i in genes_order])}\\n')\n",
    "    f.write(f'LEGEND_COLORS,{\",\".join([gene2color[i] for i in genes_order])}\\n')\n",
    "    \n",
    "    f.write(f'FIELD_LABELS,{\",\".join([i for i in genes_order])}\\n')\n",
    "    f.write(f'FIELD_COLORS,{\",\".join([gene2color[i] for i in genes_order])}\\nDATA\\n')\n",
    "    binary_data = crosstab.reset_index(drop=False).astype(str).\\\n",
    "    agg(','.join, axis=1).values\n",
    "    for i in binary_data:\n",
    "        f.write(i+'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9351a188-307f-49e2-805b-87b3df459491",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
